### ISSUE TITLE
Refactor overfit test

### ISSUE THREAD

**Issue by nd-02110114** [deepchem/deepchem#2388]
## 🚀 Refactor 

I think the overfit test is that it is enough to confirm training loss keep reducing. 
I think this style is more stable and faster than previous one. 

New test
```python
loss = []
# don't require many epochs 
model.fit(dataset, nb_epoch=10, all_losses=loss)
# new assertion
assert loss == sorted(loss, reverse=True)
```

Previous test
```python
model.fit(dataset, nb_epoch=100)
# score assertion
scores = model.evaluate(dataset, [classification_metric])
assert scores[classification_metric.name] > 0.9
```

## Motivation

Our test is really slow and flaky. These situation are blocking our rapid development. 
I seem that the bottle neck is overfit tests. 

The problem of overfit tests is confirming whether to overcome the manual threshold. There is no correct definition for how to set the threshold and some models need many epochs to overcome the threshold. 
I think we need more stable and faster test for upgrading libraries more rapidly. (Currently, we need to upgrade Numpy >=1.20, Tensorflow >=2.4, Pip >=20.3 ....)

## Additional context

Nothing

**Comment by rbharath** [comment]
This is a really interesting suggestion! I think this is pretty reasonable on my end. The only issue I can see is that sometimes there can be natural variability in the training loss, but I'm not sure if that would be an issue on a small synthetic dataset. I'd be pretty open to us swapping over our tests to this style

**Comment by peastman** [comment]
I'm not convinced this is really testing the same thing.  The point of the overfit tests is that the model should be able to achieve very low error on a tiny dataset.  Checking for the loss to decrease doesn't test that.  All it really checks is that the optimizer is functioning as an optimizer: it's causing the loss to decrease.  But the loss may still be very high.  This doesn't test that the model is capable of fitting the data well.

**Comment by nd-02110114** [comment]
> The only issue I can see is that sometimes there can be natural variability in the training loss, but I'm not sure if that would be an issue on a small synthetic dataset. 

I'm also concerned about this point. Considering this point, new tests may be sometimes flaky tests... (I assume that we overcome this point if we fix the seed value)

> I'm not convinced this is really testing the same thing. The point of the overfit tests is that the model should be able to achieve very low error on a tiny dataset. Checking for the loss to decrease doesn't test that. All it really checks is that the optimizer is functioning as an optimizer: it's causing the loss to decrease. But the loss may still be very high. This doesn't test that the model is capable of fitting the data well.

Yes. The new test is not same as the previous one. But, I'm really wondering this point, "the model should be able to achieve very low error on a tiny dataset." If we train tiny dataset, we often meet not only overfitting but also underfitting. Actually, many torch model couldn't achieve the threshold which is used previously. (The model itself has no issues.) Given that we changes the threshold for each models, I think it is not meaningful to meet the threshold with many epochs. Considering about underfitting, can we define the threshold correctly? Instead of the current tests, I think it is reasonable that we make contributors submit scores for QM9 or other dataset when they add new models.

**Comment by peastman** [comment]
I don't think underfitting is an issue.  Most of these models are universal approximators.  As long as the number of model parameters is large compared to the number of data points, they should be able to reproduce the training data to very high accuracy.  Different tests use different tolerances, but it's always very small.  It's usually limited by the number of training epochs we're willing to run, not the model's ability to reproduce the data.

**Comment by nd-02110114** [comment]
I also don't think the main issue is not underfitting. 
The problem of overfit tests is two points.

**It is difficult to set correct tolerances**

Can we define tolerances correctly? My answer is no. You mentioned, "they should be able to reproduce the training data to very high accuracy" and "the model should be able to achieve very low error on a tiny dataset." This is subjective. Can you show the concrete criteria like "roc-auc > 0.9" or "MAE<0.5" to judge that a model is capable of fitting the data well? I want to know.

> Different tests use different tolerances, but it's always very small. 

I don't think so. GCNModel test use MAE < 0.5, but GraphCovModel test use MAE < 0.1. I seem that this is not small difference. 

**Some models requires much epochs to achieve the tolerances**

I know theoretically DNN models are universal approximators. But, some models require much epochs (almost 30s ~60s) to reproduce the training data to very high accuracy. Currently, Given that models will be added in the future, the overfit tests will be blocking our CI jobs. Is this OK for our developments?


Rather than continuing with unclear and time-consuming tests,
I believe it would be better for the future if we adopted clearer and faster tests.

If we want to confirm that the model is capable of fitting the data well, I think it is reasonable that we make contributors submit scores for QM9 or other dataset when they add new models.

**Comment by peastman** [comment]
> Can you show the concrete criteria like "roc-auc > 0.9" or "MAE<0.5" to judge that a model is capable of fitting the data well?

A sufficiently large model trained for a sufficient number of epochs should be able to fit it to arbitrary accuracy.  But of course, we're also concerned about time, so we don't use huge models trained for a huge number of epochs.  We try to come up with a model size and number of epochs that lets us get the best fit possible in a reasonable amount of time.  For more expensive models, we may have to settle for a larger tolerance because we want to keep the test fast.

Consider `test_graph_conv_regression_model()`.  The data we're fitting to is normally distributed random numbers with mean 0, variance 1.  That means typical values are on the order of about +/- 1.  If the MAE <<1, we're basically doing a good job of learning the data.  The errors in predictions are small compared to the differences between values.  Given that, the tolerance of 0.1 makes sense.

Here are the actual losses (L2 loss) every 10 epochs from a run of the test:

> [0.8981979370117188, 0.37907772064208983, 0.24303672313690186, 0.1647523283958435, 0.11260946989059448, 0.08038564920425414, 0.05182197093963623, 0.032632336020469666, 0.021565908193588258, 0.014248155057430267, 0.009354925155639649, 0.005113423615694046, 0.00324598029255867, 0.0016376446932554246, 0.0008192278444766998, 0.00043272399343550203, 0.00021788235753774642, 0.00011032965267077088, 6.78852607961744e-05, 3.343723656143993e-05]

After the first 10 epochs, the MAE is on the order of 1, similar to the differences between datapoints.  Clearly that hasn't learned the data yet.  By the end of training, it's down to 3.3e-5, and the MAE is 0.0078.  That's much less than 1, and in fact much less than our tolerance of 0.1.  Perhaps we could use a lower tolerance, but that might increase the risk of occasional failures.

Now suppose the sequence of losses were something like this:

> [0.935, 0.921, 0.918. ..., 0.902, 0.901]

That would clearly indicate the model wasn't learning the data.  Even at the end of training, the error in each datapoint would be similar to the differences between datapoints.  Yet the test you proposed would pass.  The losses decrease monotonically.

**Comment by nd-02110114** [comment]
> We try to come up with a model size and number of epochs that lets us get the best fit possible in a reasonable amount of time. For more expensive models, we may have to settle for a larger tolerance because we want to keep the test fast.

Is there any concrete idea? The GNN models implemented by PyTorch has already faced this problem.

> That would clearly indicate the model wasn't learning the data. Even at the end of training, the error in each datapoint would be similar to the differences between datapoints. Yet the test you proposed would pass. The losses decrease monotonically.

In the case that we set a larger tolerance like "MAE < 0.5" to keep the test fast, I seem that the same situation happens. The model may just be slow to train. The previous overfit test is valid only when the model converges in a few tens of epochs (10~50 epochs). 

Therefore, If we want to confirm that the model is really capable of fitting the data well, I think it is reasonable that we make contributors submit scores for QM9 or other dataset when they add new models.

**Comment by peastman** [comment]
> Actually, the GAT model implemented by PyTorch often show a large MAE like 0.802 with 100 epochs in the regression tests, but don't have any problems.

We should see if we can improve that by tuning the test a bit better.  Overfitting a tiny dataset should be a very easy learning problem.  Possibly we just need to adjust a few parameters like the model size and learning rate.

> Therefore, if we want to confirm that the model is really capable of fitting the data well, I think it is reasonable that we make contributors submit scores for QM9 or other dataset when they add new models.

Definitely!  The unit tests are to catch if a code change breaks the model.  They aren't a substitute for demonstrating performance on real world datasets.

**Comment by nd-02110114** [comment]
> We should see if we can improve that by tuning the test a bit better. Overfitting a tiny dataset should be a very easy learning problem. Possibly we just need to adjust a few parameters like the model size and learning rate.

I have already tuned learning rate, the number of attention heads, the number of convolution layers and so on. The MAE value varies between 0.4 and 0.8. 

I seem there is no concrete approaches about setting a tolerance considering about a model size and number of epochs. This is a main issue of overfit tests. And, I feel there is no clear answer from you about this issue.

**Comment by peastman** [comment]
> And, I feel there is no clear answer from you about this issue.

The answer is very simple: the tolerance needs to be small compared to the typical differences between `y` values.  If the data is normally distributed with a variance of 1, a tolerance of 0.1 is reasonable.  A tolerance of 0.5 is not.  We need to fix that test in some way.  If the only way to do it is to run a lot more epochs, we'll just need to mark the test as slow.

> My proposed tests are clearer and faster tests compared with overfit tests.

But it doesn't provide any information about the question we care about: is the model implemented correctly?  If it can overfit a tiny dataset, that's one piece of evidence suggesting it works correctly.  It's far from a complete proof, of course.  The overfit tests are just one piece of a complete testing strategy.  But they provide useful information.  In contrast, checking for decreasing losses provides no information at all about the model.  It is solely a test of the optimizer.  If you doubt that, try implementing your test for a trivial model that just outputs a fixed (learnable) vector, independent of its input.  The only thing it can learn is the mean of `y` over the training set.  That model is clearly broken, yet you will still find that the loss decreases monotically during training.

Furthermore, if you ever see the loss increase during training, that probably doesn't indicate a problem with the model.  It most likely just means the learning rate is too high.

**Comment by nd-02110114** [comment]
> The answer is very simple: the tolerance needs to be small compared to the typical differences between y values. If the data is normally distributed with a variance of 1, a tolerance of 0.1 is reasonable. A tolerance of 0.5 is not. We need to fix that test in some way. If the only way to do it is to run a lot more epochs, we'll just need to mark the test as slow.

This is the answer I really wanted. And, we should need to clarify this point in docs. According to this answer, most of overfit tests for GNNs implemented by PyTorch are broken and need to mark the test as slow. 

> But it doesn't provide any information about the question we care about: is the model implemented correctly? If it can overfit a tiny dataset, that's one piece of evidence suggesting it works correctly. It's far from a complete proof, of course. The overfit tests are just one piece of a complete testing strategy. But they provide useful information. In contrast, checking for decreasing losses provides no information at all about the model. It is solely a test of the optimizer. If you doubt that, try implementing your test for a trivial model that just outputs a fixed (learnable) vector, independent of its input. The only thing it can learn is the mean of y over the training set. That model is clearly broken, yet you will still find that the loss decreases monotically during training.
Furthermore, if you ever see the loss increase during training, that probably doesn't indicate a problem with the model. It most likely just means the learning rate is too high.

I agree with these points.

**Comment by nd-02110114** [comment]
I could understand your opinions, so I closed this issue.
